### ISSUE TITLE
Tree accessibility issue -- keyboard trap

### ISSUE THREAD

**Issue by Charlie243** [primefaces/primefaces#1547]
My PrimeFaces-based webapp was recently reviewed for accessibility compliance by a US Federal agency.  One issue that was tagged for non-compliance was a "keyboard trap" resulting from the inability to go "backwards" (shift + tab) from a PrimeFaces tree component.  That is, you can go forward from a tree to the next component using the tab key, but you CAN'T go backwards to a previous component using shift + tab.  This is a showstopper bug for users that depend on keyboard navigation -- like some users with visual impairments for example.  And my webapp cannot be OK'd as accessibility-compliant until this bug is fixed.

This issue was discussed in the PrimeFaces forum here: [http://forum.primefaces.org/viewtopic.php?f=3&t=46005](url)

The behavior can be seen in the selection tree examples in the PrimeFaces showcase.  But I also included a simple code example in my forum report of the issue.

The bug is in PrimeFaces 6.0 but was also in the 5.x releases (and probably earlier).  

Thanks for taking a look at this issue.  I hope it can be fixed relatively easily (and quickly).

Charlie

**Comment by tandraschko** [comment]
JFYI: if you need a quick fix, you should consider PrimeFaces PRO.

**Comment by Charlie243** [comment]
Thanks Thomas. I love PrimeFaces and wish I could go PRO. (I'm Elite.)  But I'm in the academic/non-profit world and it's just not financially feasible.  

Also, I think I messed up the link to the forum discussion.  So let me try it again:
http://forum.primefaces.org/viewtopic.php?f=3&t=46005

Cheers,
Charlie

**Comment by Charlie243** [comment]
In order to facilitate the debugging and resolution of this bug I'm including a simple code example here that shows the incorrect behavior.  (This is the same example that I included in my original post on the PrimeFaces forum.)

index.xhtml

```
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:h="http://xmlns.jcp.org/jsf/html"
      xmlns:p="http://primefaces.org/ui">
    <h:head>
        <title>Tree Navigation Bug</title>
    </h:head>
    <h:body>
        <h:form id="form">
            <p:growl id="messages" showDetail="true" escape="false" sticky="true"/>
            <p:outputLabel value="Input #1: "/>
            <p:inputText />
            <p/>

            <p:tree
                id="checkboxTree"
                value="#{treeBean.root}" 
                var="node" 
                selectionMode="checkbox"
                selection="#{treeBean.selectedNodes}">

                <p:treeNode>
                    <h:outputText value="#{node}" />
                </p:treeNode>
            </p:tree>

            <p/>
            <p:commandButton 
                value="Display Selected" 
                update="messages"
                action="#{treeBean.displaySelectedMultiple}" 
                id="btnDisplay"/>
            <p/>
            <p:outputLabel value="Input #2: "/>
            <p:inputText />
        </h:form>
    </h:body>
</html>

```

TreeBean.java

```

import java.io.Serializable;
import javax.faces.application.FacesMessage;
import javax.faces.bean.ManagedBean;
import javax.faces.bean.ViewScoped;
import javax.faces.context.FacesContext;
import org.primefaces.model.CheckboxTreeNode;
import org.primefaces.model.TreeNode;

@ManagedBean
@ViewScoped
public class TreeBean implements Serializable {

    private TreeNode root;
    private TreeNode[] selectedNodes;

    public TreeBean() {
        root = new CheckboxTreeNode("Root", null);
        TreeNode node0 = new CheckboxTreeNode("Node 0", root);
        TreeNode node1 = new CheckboxTreeNode("Node 1", root);
        TreeNode node2 = new CheckboxTreeNode("Node 2", root);

        TreeNode node00 = new CheckboxTreeNode("Node 0.0", node0);
        TreeNode node01 = new CheckboxTreeNode("Node 0.1", node0);

        TreeNode node10 = new CheckboxTreeNode("Node 1.0", node1);
        TreeNode node11 = new CheckboxTreeNode("Node 1.1", node1);

        TreeNode node000 = new CheckboxTreeNode("Node 0.0.0", node00);
        TreeNode node001 = new CheckboxTreeNode("Node 0.0.1", node00);
        TreeNode node010 = new CheckboxTreeNode("Node 0.1.0", node01);

        TreeNode node100 = new CheckboxTreeNode("Node 1.0.0", node10);
    }

    public TreeNode getRoot() {
        return root;
    }

    public TreeNode[] getSelectedNodes() {
        return selectedNodes;
    }

    public void setSelectedNodes(TreeNode[] selectedNodes) {
        this.selectedNodes = selectedNodes;
    }

    public void displaySelectedMultiple() {
        if (selectedNodes != null && selectedNodes.length > 0) {
            StringBuilder builder = new StringBuilder();

            for (TreeNode node : selectedNodes) {
                builder.append(node.getData().toString());
                builder.append("<br />");
            }

            FacesMessage message = new FacesMessage(FacesMessage.SEVERITY_INFO, "Selected", builder.toString());
            FacesContext.getCurrentInstance().addMessage(null, message);
        }
    }
}   
```

As I said on the PrimeFaces forum: the issue is basic. You can tab forward through the components -- input text box #1, checkbox tree, "Display Selected" command button, and input text box #2. But if you try to shift + tab backwards through the components you can go from input text box #2 to the "Display Selected" command button, to the checkbox tree -- but there you are "trapped". The "shift + tab" combination will only get you to the top level of the tree but will never get you back to input text box #1. 

Again, this is a showstopper bug for keyboard accessibility -- making any user interface with the PrimeFaces tree unusable for users who depend on the keyboard.

Thanks,
Charlie
